File: src/main/java/org/mercurialftc/mercurialftc/scheduler/commands/ParallelCommandGroup.java
package org.mercurialftc.mercurialftc.scheduler.commands;

import org.jetbrains.annotations.NotNull;
import org.mercurialftc.mercurialftc.scheduler.OpModeEX;
import org.mercurialftc.mercurialftc.scheduler.Scheduler;
import org.mercurialftc.mercurialftc.scheduler.subsystems.SubsystemInterface;

import java.util.*;

@SuppressWarnings("unused")
public class ParallelCommandGroup implements CommandGroup {
	private final Map<Command, Boolean> commands;
	private final Set<SubsystemInterface> requiredSubsystems;
	private final Set<OpModeEX.OpModeEXRunStates> runStates;
	private boolean interruptable;

	/**
	 * a new empty ParallelCommandGroup, which will run all its commands at the same time
	 */
	public ParallelCommandGroup() {
		this.requiredSubsystems = new HashSet<>();
		this.runStates = new HashSet<>(2);
		interruptable = true;
		this.commands = new HashMap<>();
	}

	private ParallelCommandGroup(@NotNull HashMap<Command, Boolean> commands, Set<SubsystemInterface> requirements, Set<OpModeEX.OpModeEXRunStates> runStates, boolean interruptable) {
		this.requiredSubsystems = requirements;
		this.runStates = runStates;
		this.interruptable = interruptable;
		this.commands = commands;
	}

	/**
	 * non-mutating
	 *
	 * @param commands new commands to add
	 * @return a new ParallelCommandGroup, with the added commands
	 */
	@Override
	public ParallelCommandGroup addCommands(Command... commands) {
		return addCommands(Arrays.asList(commands));
	}

	/**
	 * non-mutating
	 *
	 * @param commands new commands to add
	 * @return a new ParallelCommandGroup, with the added commands
	 */
	@Override
	public ParallelCommandGroup addCommands(Collection<Command> commands) {
		if (this.commands.containsValue(true)) {
			throw new IllegalStateException(
					"Commands cannot be added to a composition while it's running");
		}

		HashMap<Command, Boolean> newCommandMap = new HashMap<>(this.commands);

		Set<SubsystemInterface> newRequirementSet = new HashSet<>(this.getRequiredSubsystems());
		boolean newInterruptable = interruptable();
		HashSet<OpModeEX.OpModeEXRunStates> newRunStates = new HashSet<>(2);


		for (Command command : commands) {
			if (!Collections.disjoint(command.getRequiredSubsystems(), getRequiredSubsystems())) {
				throw new IllegalArgumentException(
						"Multiple commands in a parallel composition cannot require the same subsystems");
			}
			newCommandMap.put(command, false);
			newRequirementSet.addAll(command.getRequiredSubsystems());
			newInterruptable &= command.interruptable();
			newRunStates.addAll(command.getRunStates());
		}

		Scheduler.getSchedulerInstance().registerComposedCommands(commands);

		return new ParallelCommandGroup(
				newCommandMap,
				newRequirementSet,
				newRunStates,
				newInterruptable
		);
	}

	@Override
	public final boolean interruptable() {
		return interruptable;
	}

	@Override
	public final void initialise() {
		if (commands.isEmpty()) {
			throw new RuntimeException("Attempted to run empty ParallelCommandGroup, ParallelCommandGroup requires a minimum of 1 Command to be run");
		}
		interruptable = true;
		for (Map.Entry<Command, Boolean> commandRunning : commands.entrySet()) {
			if (commandRunning.getKey().getRunStates().contains(Scheduler.getSchedulerInstance().getRunState())) {
				commandRunning.getKey().initialise();
				commandRunning.setValue(true);
				interruptable &= commandRunning.getKey().interruptable();
			} else {
				commandRunning.setValue(false);
			}
		}
	}

	@Override
	public final void execute() {
		interruptable = true;
		for (Map.Entry<Command, Boolean> commandRunning : commands.entrySet()) {
			if (!(commandRunning.getValue() && commandRunning.getKey().getRunStates().contains(Scheduler.getSchedulerInstance().getRunState()))) {
				continue;
			}
			Command command = commandRunning.getKey();

			if (command.finished()) {
				command.end(false);
				commandRunning.setValue(false);
			} else {
				interruptable &= command.interruptable();
				command.execute();
			}
		}
	}

	@Override
	public final void end(boolean interrupted) {
		for (Map.Entry<Command, Boolean> commandRunning : commands.entrySet()) {
			if (commandRunning.getValue()) {
				Command command = commandRunning.getKey();
				if (command.getRunStates().contains(Scheduler.getSchedulerInstance().getRunState())) {
					command.end(interrupted);
				}
			}
		}
	}

	@Override
	public final Set<SubsystemInterface> getRequiredSubsystems() {
		return requiredSubsystems;
	}

	@Override
	public final Set<OpModeEX.OpModeEXRunStates> getRunStates() {
		return runStates;
	}

	@Override
	public final boolean finished() {
		return !commands.containsValue(true);
	}
}

File: Mercurial/src/main/kotlin/dev/frozenmilk/mercurial/commands/ParallelCommandGroup.kt
package dev.frozenmilk.mercurial.commands

import dev.frozenmilk.mercurial.Mercurial
import dev.frozenmilk.mercurial.RunStates
import dev.frozenmilk.mercurial.subsystems.Subsystem
import java.util.Arrays
import java.util.Collections

@Suppress("unused")
class ParallelCommandGroup : CommandGroup {
	private val commands: MutableMap<Command, Boolean>
	override val requiredSubsystems: Set<Subsystem>
	override val runStates: Set<RunStates>
	private var _interruptible: Boolean

	/**
	 * a new empty ParallelCommandGroup, which will run all its commands at the same time
	 */
	constructor() {
		requiredSubsystems = HashSet()
		runStates = HashSet(2)
		_interruptible = true
		commands = HashMap()
	}

	private constructor(commands: HashMap<Command, Boolean>, requirements: Set<Subsystem>, runStates: Set<RunStates>, interruptable: Boolean) {
		requiredSubsystems = requirements
		this.runStates = runStates
		this._interruptible = interruptable
		this.commands = commands
	}

	/**
	 * non-mutating
	 *
	 * @param commands new commands to add
	 * @return a new ParallelCommandGroup, with the added commands
	 */
	override fun addCommands(vararg commands: Command): ParallelCommandGroup {
		return addCommands(Arrays.asList(*commands))
	}

	/**
	 * non-mutating
	 *
	 * @param commands new commands to add
	 * @return a new ParallelCommandGroup, with the added commands
	 */
	override fun addCommands(commands: Collection<Command>): ParallelCommandGroup {
		check(!this.commands.containsValue(true)) { "Commands cannot be added to a composition while it's running" }
		val newCommandMap = HashMap(this.commands)
		val newRequirementSet = HashSet(requiredSubsystems)
		var newInterruptable = interruptible
		val newRunStates = HashSet<RunStates>(2)
		for (command in commands) {
			require(Collections.disjoint(command.requiredSubsystems, requiredSubsystems)) { "Multiple commands in a parallel composition cannot require the same subsystems" }
			newCommandMap[command] = false
			newRequirementSet.addAll(command.requiredSubsystems)
			newInterruptable = newInterruptable and command.interruptible
			newRunStates.addAll(command.runStates)
		}
		Mercurial.registerComposedCommands(commands)
		return ParallelCommandGroup(
				newCommandMap,
				newRequirementSet,
				newRunStates,
				newInterruptable
		)
	}

	override val interruptible: Boolean
		get() {
			return _interruptible
		}

	override fun initialise() {
		if (commands.isEmpty()) {
			throw RuntimeException("Attempted to run empty ParallelCommandGroup, ParallelCommandGroup requires a minimum of 1 Command to be run")
		}
		_interruptible = true
		for (commandRunning in commands.entries) {
			if (commandRunning.key.runStates.contains(Mercurial.runState)) {
				commandRunning.key.initialise()
				commandRunning.setValue(true)
				_interruptible = _interruptible and commandRunning.key.interruptible
			} else {
				commandRunning.setValue(false)
			}
		}
	}

	override fun execute() {
		_interruptible = true
		for (commandRunning in commands.entries) {
			if (!(commandRunning.value && commandRunning.key.runStates.contains(Mercurial.runState))) {
				continue
			}
			val command = commandRunning.key
			if (command.finished()) {
				command.end(false)
				commands[command] = false
			} else {
				_interruptible = _interruptible and command.interruptible
				command.execute()
			}
		}
	}

	override fun end(interrupted: Boolean) {
		for ((command, value) in commands) {
			if (value) {
				if (command.runStates.contains(Mercurial.runState)) {
					command.end(interrupted)
				}
			}
		}
	}

	override fun finished(): Boolean {
		return !commands.containsValue(true)
	}
}
